add support to stylelint 16 bro

Please support stylelint 16.0.0, the big one

Now the code is in ESM instead of CommonJS 

Very good

Stylelint team published docs to know how to migrate with ease.

No need to keep support to commons from now on.

Style lint 16 is not allowed in your package.json. remove this restriction to

@AndyOGo Hi, firstly, thanks for your `stylelint` plugin! Great job here, we use it in our `stylelint` config at GitLab. We are slowly migrating our configuration to the `stylelint@16` and this plugin is the last dependency we need to update. I would like to help with that migration. I briefly looked at the repo and I think there is no straightforward migration, so before doing anything I want to share my plan with you to confirm that it looks good.

1. One of the dependencies - https://github.com/stylelint/jest-preset-stylelint pinned the minimum required Node.js version to v18. Do you think this plugin should do the same?
2. Additionally, the mentioned dep `jest-preset-stylelint` requires an upgrade to v7 to support `stylelint@16`.
3. I tried to build the repo and got the following error:

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: stylelint-declaration-strict-value@1.9.2
npm ERR! Found: typedoc@0.25.4
npm ERR! node_modules/typedoc
npm ERR!   dev typedoc@"^0.22.7" from the root project
npm ERR!   peer typedoc@">=0.24.0" from typedoc-plugin-markdown@3.17.1
npm ERR!   node_modules/typedoc-plugin-markdown
npm ERR!     dev typedoc-plugin-markdown@"^3.2.1" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! dev typedoc@"^0.22.7" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: typescript@4.7.4
npm ERR! node_modules/typescript
npm ERR!   peer typescript@"4.0.x || 4.1.x || 4.2.x || 4.3.x || 4.4.x || 4.5.x || 4.6.x || 4.7.x" from typedoc@0.22.18
npm ERR!   node_modules/typedoc
npm ERR!     dev typedoc@"^0.22.7" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

Seems like we need to update the `typedoc` dependency. Is that okay?

4. After updating the `typedoc` dependency I got another error:

```
❯ npm run build

> stylelint-declaration-strict-value@1.9.2 build
> microbundle --tsconfig tsconfig.build.json --compress

(rpt2 plugin) TypeError: Cannot read properties of undefined (reading 'length')
TypeError: Cannot read properties of undefined (reading 'length')
    at getPropertyOfType (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:44635:45)
    at getExternalModuleMember (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:37183:46)
    at getTargetOfImportSpecifier (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:37267:28)
    at getTargetOfAliasDeclaration (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:37325:28)
    at resolveAlias (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:37364:30)
    at tryResolveAlias (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:37380:24)
    at getCandidateName (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:57441:33)
    at Object.getSpellingSuggestion (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:1883:33)
    at getSpellingSuggestionForName (/Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:57431:23)
    at /Users/vslobodin/Development/stylelint-declaration-strict-value/node_modules/microbundle/node_modules/typescript/lib/typescript.js:57376:34
```

Not sure what this error is about.

These just are some things that probably could be addressed, or probably not. Any way, before doing anything, I would like to get your thoughts. Thanks!

--- a/jest.setup.js
+++ b/jest.setup.js
@@ -1,8 +1,11 @@
-/* eslint-disable @typescript-eslint/no-var-requires */
-const { getTestRule } = require('jest-preset-stylelint');
-const { lint } = require('stylelint');
+import { getTestRule } from 'jest-preset-stylelint';
+import { lint } from 'stylelint';
+// eslint-disable-next-line import/extensions
+import declarationStrictValuePlugin from './src/index.ts';
 
-global.testRule = getTestRule({ plugins: ['./src/index.ts'] });
+const plugins = [declarationStrictValuePlugin];
+
+global.testRule = getTestRule({ plugins });
 
 global.testCustomAutoFixMessage = testCustomAutoFixMessage;
 
@@ -19,11 +22,12 @@ function testCustomAutoFixMessage({ ruleName, config, reject, fix }) {
             fix,
             code,
             config: {
-              plugins: ['./src/index.ts'],
+              plugins,
               rules: {
                 [ruleName]: config,
               },
             },
+            quietDeprecationWarnings: true,
           });
 
           return { message, warnings };

--- a/lint-staged.config.js
+++ b/lint-staged.config.js
@@ -1,7 +1,7 @@
 module.exports = {
   '(src|test)/**/*.[tj]s': (filenames) => [
     `eslint --fix ${filenames.join(' ')}`,
-    'jest',
+    'cross-env NODE_OPTIONS="--experimental-vm-modules --no-warnings" jest',
   ],
   '(README).md': ["doctoc --title '**Table of Contents**'"],
 };

--- a/package.json
+++ b/package.json
@@ -3,14 +3,22 @@
   "version": "1.9.2",
   "description": "Specify properties for which a variable, function, keyword or value must be used",
   "source": "src/index.ts",
+  "exports": {
+    "types": "./dist/index.d.ts",
+    "require": "./dist/index.js",
+    "default": "./dist/index.mjs.js"
+  },
   "main": "dist/index.js",
   "module": "dist/index.esm.js",
   "unpkg": "dist/index.umd.js",
   "types": "dist/index.d.ts",
+  "engines": {
+    "node": ">=18.12.0"
+  },
   "scripts": {
     "build": "microbundle --tsconfig tsconfig.build.json --compress",
     "release": "dotenv semantic-release",
-    "test": "jest",
+    "test": "cross-env NODE_OPTIONS=\"--experimental-vm-modules --no-warnings\" jest",
     "coverage": "jest --config='./jest.coverage.config.js'",
     "postcoverage": "dotenv codecov",
     "eslint": "eslint 'src/**/*.[tj]s' 'test/**/*.[tj]s'",
@@ -45,7 +53,7 @@
   },
   "homepage": "https://github.com/AndyOGo/stylelint-declaration-strict-value#readme",
   "peerDependencies": {
-    "stylelint": ">=7 <=15"
+    "stylelint": ">=7 <=16"
   },
   "devDependencies": {
     "@babel/cli": "^7.12.8",
@@ -68,6 +76,7 @@
     "babel-jest": "^29.4.2",
     "babel-register-ts": "^7.0.0",
     "codecov": "^3.8.1",
+    "cross-env": "^7.0.3",
     "doctoc": "^2.0.0",
     "dotenv-cli": "^4.0.0",
     "eslint": "^7.14.0",
@@ -78,14 +87,14 @@
     "eslint-plugin-tsdoc": "^0.2.7",
     "husky": "^6.0.0",
     "jest": "^29.4.2",
-    "jest-preset-stylelint": "^6.1.0",
+    "jest-preset-stylelint": "^7.0.0",
     "lint-staged": "^10.5.2",
-    "microbundle": "^0.12.0",
+    "microbundle": "^0.15.1",
     "nyc": "^15.1.0",
     "pinst": "^2.1.6",
     "prettier": "^2.2.1",
     "semantic-release": "^17.3.0",
-    "stylelint": "^15.1.0",
+    "stylelint": "^16.1.0",
     "typedoc": "^0.22.7",
     "typedoc-plugin-markdown": "^3.2.1",
     "typescript": "^4.1.2"

--- a/src/defaults.ts
+++ b/src/defaults.ts
@@ -1,4 +1,4 @@
-import type { Node, Root } from 'stylelint/node_modules/postcss';
+import type { Node, Root } from 'postcss';
 
 /**
  * Rule Name.

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,4 +1,4 @@
-import type { Declaration, Root, AtRule } from 'stylelint/node_modules/postcss';
+import type { Declaration, Root, AtRule } from 'postcss';
 import stylelint, { PostcssResult, Rule } from 'stylelint';
 import shortCSS from 'shortcss';
 import list from 'shortcss/lib/list';

